src/imitation/regularization/__init__.py

```diff
@@ -0,0 +1,254 @@
+                penalty += th.sum(param.data.pow(2))
+        return self.lambda_ * penalty
+
+
+class L1Regularizer(LossRegularizer):
+    """Applies L1 regularization to a loss function."""
+
+    def _regularize_loss(self, loss: th.Tensor) -> Union[float, th.Tensor]:
+        """Returns the loss penalty.
+
+        Calculates the L1 norm of the weights in the optimizer,
+        and returns a scaled version of it as the penalty.
+
+        Args:
+            loss: The loss to regularize.
+
+        Returns:
+            The scaled L1 norm of the weights.
+        """
+        penalty = 0
+        for group in self.optimizer.param_groups:
+            for param in group["params"]:
+                penalty += th.sum(th.abs(param.data))
+        return self.lambda_ * penalty
+
+
+class WeightDecayRegularizer(WeightRegularizer):
+    """Applies weight decay to a loss function."""
+
+    def _regularize_weight(self, weight, group) -> Union[float, th.Tensor]:
+        """Returns the weight penalty.
+
+        Args:
+            weight: The weight to regularize.
+            group: The group of parameters to which the weight belongs.
+
+        Returns:
+            The weight penalty (to add to the current value of the weight)
+        """
+        return -self.lambda_ * group["lr"] * weight.data
+
+
+class IntervalParamScaler:
+    """Scales the lambda of the regularizer by some constant factor.
+
+    Lambda is scaled up if the ratio of the validation loss to the training lossq
+    is above the tolerable interval, and scaled down if the ratio is below the
+    tolerable interval. Nothing happens if the ratio is within the tolerable
+    interval.
+    """
+
+    def __init__(self, scaling_factor: float, tolerable_interval: Tuple[float, float]):
+        """Initialize the interval parameter scaler.
+
+        Args:
+            scaling_factor: The factor by which to scale the lambda.
+            tolerable_interval: The interval within which the ratio of the validation
+                loss to the training loss is considered acceptable.
+        """
+        self.scaling_factor = scaling_factor
+        self.tolerable_interval = tolerable_interval
+
+    def __call__(
+        self,
+        lambda_: float,
+        train_loss: th.Tensor,
+        val_loss: th.Tensor,
+    ) -> float:
+        """Scales the lambda of the regularizer by some constant factor.
+
+        Lambda is scaled up if the ratio of the validation loss to the training loss
+        is above the tolerable interval, and scaled down if the ratio is below the
+        tolerable interval. Nothing happens if the ratio is within the tolerable
+        interval.
+
+        Args:
+            lambda_: The current value of the lambda.
+            train_loss: The loss on the training set.
+            val_loss: The loss on the validation set.
+
+        Returns:
+            The new value of the lambda.
+        """
+        # TODO(juan) does this division below work? they are tensors.
```

Comment:
Adding an assert that they're scalars (or at least are scalar after squeezing) seems good.

Reply:
wdym after squeezing?

Reply:
Also, how do you actually check the rank of a tensor. Google suggests dimension but dimension is mathematically not the same as the rank, even though this might be the folk term for it in the ML community.

Reply:
By squeezing I mean `th.squeeze` -- removing any length-1 dimensions. So we might want to treat a tensor of shape `(1,)` as a scalar. (Maybe -- depends on the context.)

`dim` is the number of indices needed to index the tensor, which I think lines up with the mathematical notion of rank (as opposed to the dimension of the space it's embedded in), but I'm rusty on tensors in maths.